File: bms/go/geo/geo_s1_Interval_ContainsInterval/geo_s1_Interval_ContainsInterval.go
package test

import (
	"math"
)

func (i geo_s1_Interval) ContainsInterval(oi geo_s1_Interval) bool {
	if i.IsInverted() {
		if oi.IsInverted() {
			return oi.Lo >= i.Lo && oi.Hi <= i.Hi
		}
		return (oi.Lo >= i.Lo || oi.Hi <= i.Hi) && !i.IsEmpty()
	}
	if oi.IsInverted() {
		return i.IsFull() || oi.IsEmpty()
	}
	return oi.Lo >= i.Lo && oi.Hi <= i.Hi
}

func (i geo_s1_Interval) IsInverted() bool { return i.Lo > i.Hi }

func (i geo_s1_Interval) IsEmpty() bool { return i.Lo == math.Pi && i.Hi == -math.Pi }

func (i geo_s1_Interval) IsFull() bool { return i.Lo == -math.Pi && i.Hi == math.Pi }

type geo_s1_Interval struct {
	Lo, Hi float64
}

File: translations_only/success/go/geo/geo_s1_Interval_ContainsInterval/gpt4/Hinted/geo_s1_Interval_ContainsInterval.rs

use std::f64::consts::PI;

struct GeoS1Interval {
    lo: f64,
    hi: f64,
}

fn contains_interval(i: &GeoS1Interval, oi: &GeoS1Interval) -> bool {
    if is_inverted(i) {
        if is_inverted(oi) {
            return oi.lo >= i.lo && oi.hi <= i.hi;
        }
        return (oi.lo >= i.lo || oi.hi <= i.hi) && !is_empty(i);
    }
    if is_inverted(oi) {
        return is_full(i) || is_empty(oi);
    }
    oi.lo >= i.lo && oi.hi <= i.hi
}

fn is_inverted(i: &GeoS1Interval) -> bool {
    i.lo > i.hi
}

fn is_empty(i: &GeoS1Interval) -> bool {
    i.lo == PI && i.hi == -PI
}

fn is_full(i: &GeoS1Interval) -> bool {
    i.lo == -PI && i.hi == PI
}
